libqtile/command/base.py

```diff
@@ -55,11 +57,114 @@ class CommandException(Exception):
 class CommandObject(metaclass=abc.ABCMeta):
     """Base class for objects that expose commands
 
-    Each command should be a method named `cmd_X`, where X is the command name.
+    Any command to be exposed should be decorated with
+    `@CommandObject.expose()` (classes that are not explicitly
+    inheriting from CommandObject will need to import the module)
     A CommandObject should also implement `._items()` and `._select()` methods
     (c.f. docstring for `.items()` and `.select()`).
     """
 
+    def __new__(cls, *args, **kwargs):
+
+        # Check which level of command object has been parsed
+        # This test ensures inherited classes don't stop additional
+        # methods from being exposed
+        if getattr(cls, "_command_object", "") == cls.__name__:
+            return super(CommandObject, cls).__new__(cls)
+
+        commands = {}
+        cmd_s = set()
+
+        # We need to iterate over the class's inherited classes in reverse order
+        # We reverse the order so the exposed command will always be the latest
+        # definition of the method.
+        for c in reversed(list(cls.__mro__)):
+            base_classes = []
+
+            for method_name in list(c.__dict__.keys()):
+                method = getattr(c, method_name, None)
+
+                if method is None:
+                    continue
+
+                # Ignore methods that were defined by subclasses
+                for b in base_classes:
+                    if method == getattr(b, method_name, None):
+                        continue
+
+                # If the command has been exposed, add it to our dictionary
+                # If the method name is already in our dictionary then bind the
+                # latest definition to that command
+                if hasattr(method, "_cmd") or method_name in commands:
+                    commands[method_name] = method
+                # For now, we'll accept the old format `cmd_` naming scheme for
+                # exposing commands.
+                # NOTE: This will be deprecated in the future
+                elif method_name.startswith("cmd_"):
+                    cmd_s.add(method_name)
+                    commands[method_name[4:]] = method
+
+                # Expose additional names
+                for mapping in getattr(method, "_mapping", list()):
+                    setattr(cls, mapping, method)
+                    commands[mapping] = method
+
+                base_classes.append(method_name)
+
+        if cmd_s:
+            names = " ".join(cmd_s)
+            msg = (
+                f"The use of the 'cmd_' prefix to expose commands via IPC "
+                f"is deprecated. Methods should use the "
+                f"@CommandObject.expose() decorator instead. "
+                f"Please update: {names}"
+            )
+            logger.warning(f"Deprecation Warning: {msg}")
+
+        # Record the object as being parsed.
+        cls._command_object = cls.__name__
```

Comment:
Does this make the parsing only happen once per class definition per Qtile instance?

Reply:
That's the idea, yes. So, once we've parsed, say, `widget.TextBox` any other instantiations of the same class will not be parsed. But, if we create a widget that subclasses `TextBox` it will be parsed.